pkg/vault/contracts/Vault.sol

```diff
@@ -1336,23 +1344,24 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
                 // EXACT_IN requires the exact amount of wrapped tokens to be unwrapped, so we call redeem.
                 // The amount of wrapped tokens to redeem is the amount necessary to fulfill the trade
                 // (amountInWrapped), plus the amount needed to leave the buffer rebalanced 50/50 at the end
-                // (bufferWrappedSurplus).
-                vaultWrappedDeltaHint = amountInWrapped + bufferWrappedSurplus;
+                // (bufferWrappedToRebalance). `bufferWrappedToRebalance` may be positive if buffer has an excess of
+                // wrapped, or negative if the buffer has an excess of underlying tokens. `vaultWrappedDeltaHint`
+                // will always be a positive number, because if `abs(bufferWrappedToRebalance) > amountInWrapped` and
+                // `bufferWrappedToRebalance < 0`, it means the buffer has enough liquidity to fulfill the trade (i.e.
+                // `bufferBalances.getBalanceRaw() >= amountOutUnderlying` is true).
+                vaultWrappedDeltaHint = (amountInWrapped.toInt256() + bufferWrappedToRebalance).toUint256();
                 vaultUnderlyingDeltaHint = wrappedToken.redeem(vaultWrappedDeltaHint, address(this), address(this));
             } else {
                 // EXACT_OUT requires the exact amount of underlying tokens to be returned, so we call withdraw.
-                // The amount of underlying tokens to withdraw is the amount necessary to fulfill the trade
-                // (amountOutUnderlying), plus the amount needed to leave the buffer rebalanced 50/50 at the end
-                // (bufferUnderlyingSurplus).
-                if (bufferWrappedSurplus > 0) {
-                    // An unwrap operation withdraws wrapped tokens from the buffer and returns underlying tokens.
-                    // For the buffer to rebalance during an unwrap operation, it must have an excess (surplus) of
-                    // wrapped tokens. Since this is an EXACT_OUT unwrap (withdraw), we need to calculate the amount of
-                    // underlying tokens the buffer should receive to rebalance, based on the wrapped token surplus.
-                    // The redeem function is used here, as it performs the inverse of a withdraw operation.
-                    bufferUnderlyingSurplus = wrappedToken.previewRedeem(bufferWrappedSurplus);
+                // The amount of underlying tokens to withdraw is calculated in terms of the wra.
+                if (bufferWrappedToRebalance != 0) {
+                    vaultWrappedDeltaHint = (amountInWrapped.toInt256() + bufferWrappedToRebalance).toUint256();
+                    vaultUnderlyingDeltaHint = wrappedToken.previewRedeem(vaultWrappedDeltaHint);
+                } else {
+                    // If the buffer does not have a surplus, unwrap the exact `amountOutUnderlying` amount.
+                    vaultUnderlyingDeltaHint = amountOutUnderlying;
```

Comment:
No assignment for `vaultWrappedDeltaHint` in this case?

I get that it's assigned below unconditionally. But maybe it should be symmetrical with `_wrap`.

Reply:
unwrap and wrap are assymetrical, due to the approval. If we assigned vaultWrappedDeltaHint here, it wouldn't be used anywhere (differently from the wrap function, where we need both values).

Reply:
See related comment below (https://github.com/balancer/balancer-v3-monorepo/pull/975#discussion_r1764175974). It is asymmetric, but assigning the hints at the top (while in one case it wouldn't be used), might be a bit cleaner, and avoids the "else" clauses.

Reply:
I made it a bit more symmetrical, following @EndymionJkb suggestion.

Reply:
@jubeira @EndymionJkb , I needed to mess the symmetry of wrap and unwrap, but gained 80 bytes and, in the real world, potentially 10k of gas by removing a previewRedeem from unwrap ExactOut.